main.nf

```diff
@@ -0,0 +1,581 @@
+}
+
+
+/********************************************************************
+  process preprocessFASTA
+  Collects all the files related to the reference genome, like
+  .fai,.gz ...
+  If the user gives them as an input, they are used
+  If not they are produced in this process given only the fasta file.
+********************************************************************/
+
+
+process preprocessFASTA{
+
+  publishDir "$baseDir/sampleDerivatives"
+
+
+  input:
+  file fasta from fasta
+  file fai from fai
+  file fastagz from fastagz
+  file gzfai from gzfai
+  file gzi from gzi
+  output:
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi") into fastaChannel
+  script:
+  """
+  [[ "${params.fai}"=="nofai" ]] &&  samtools faidx $fasta || echo " fai file of user is used, not created"
+  [[ "${params.fastagz}"=="nofastagz" ]]  && bgzip -c ${fasta} > ${fasta}.gz || echo "fasta.gz file of user is used, not created "
+  [[ "${params.gzi}"=="nogzi" ]] && bgzip -c -i ${fasta} > ${fasta}.gz || echo "gzi file of user is used, not created"
+  [[ "${params.gzfai}"=="nogzfai" ]] && samtools faidx "${fasta}.gz" || echo "gz.fai file of user is used, not created"
+  """
+
+}
+
+
+/********************************************************************
+  process preprocessBAM
+  If the user gives the index files for the bam files as an input, they are used
+  If not they are produced in this process given only the fasta file.
+  Moreover this takes care of the read group line too.
+********************************************************************/
+
+
+process preprocessBAM{
+
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+  script:
+  """
+	  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { java -jar /picard.jar AddOrReplaceReadGroups \
+    I=${bam[0]} \
+    O=ready/${bam[0]} \
+    RGID=${params.rgid} \
+    RGLB=${params.rglb} \
+    RGPL=${params.rgpl} \
+    RGPU=${params.rgpu} \
+    RGSM=${params.rgsm};}
+    cd ready ;samtools index ${bam[0]};
+  """
+}
+
+
+
+fastaChannel.map{file -> tuple (1,file[0],file[1],file[2],file[3],file[4])}
+            .set{all_fa};
+
+completeChannel.map { file -> tuple(1,file[0],file[1]) }
+               .set{all_bam};
+
+all_fa.cross(all_bam)
+      .set{all_fa_bam};
+
+
+
+      /********************************************************************
+        process makeExamples
+        Getting bam files and converting them to images ( named examples )
+
+	Can be parallelized through the params.n_shards
+	( if params.n_shards >= 1 parallelization happens automatically)
+      ********************************************************************/
+
+if(params.bed){
+  process makeExamples_with_bed{
+
+      tag "${bam[1]}"
+    cpus params.j
```

Comment:
Please always specify requirements in config files so that they can be overwritten / removed.